[FLOC-1824] Get the local instance ID by comparing IP server addresses

wallrj opened the pull request:
Fixes https://clusterhq.atlassian.net/browse/FLOC-1824

The interesting build results are in the `expected_failures` section and not listed below. Here's a direct link:
 * http://build.clusterhq.com/builders/flocker%2Ffunctional%2Frackspace%2Fcentos-7%2Fstorage-driver/builds/1

wallrj commented on flocker/node/agents/cinder.py:
```diff
@@ -169,36 +180,60 @@ def wait_for_volume(volume_manager, expected_volume,
         """
-        Look up the Xen instance ID for this node.
+        Find the Nova API server with the same non-loopback IP
+        addresses as this node.
```
Well, in fact we now look for the API server addresses that are a subset of the locally discovered IPv4 and IPv6 addresses.
Probably should update this docstring.

wallrj commented on flocker/node/agents/cinder.py:
```diff
@@ -169,36 +180,60 @@ def wait_for_volume(volume_manager, expected_volume,
-        command = [b"xenstore-read", b"name"]
-        return check_output(command).strip().decode("ascii")[len(prefix):]
+        local_ips = get_all_ips()
+        for server in self.nova_server_manager.list(detailed=True):
```
The `detailed=True` argument is the default. And I didn't include it in `INovaServerManager` so I should probably remove it here.

wallrj commented on flocker/node/agents/cinder.py:
```diff
@@ -305,8 +340,8 @@ def detach_volume(self, blockdevice_id):
         wait_for_volume(
-            volume_manager=self.nova_volume_manager,
-            expected_volume=nova_volume,
+            volume_manager=self.cinder_volume_manager,
```
Switching to `cinder_api` instead of instantiating `CinderBlockAPI` revealed that `wait_for_volume` was calling `list` which isn't declared in the `INovaVolumeManager` interface.
So I now use the cinder client to get the volume and to wait for it's status to change.

exarkun replied:
Ah, makes sense.

wallrj commented on flocker/node/agents/functional/test_cinder.py:
```diff
@@ -79,16 +79,16 @@ def test_foreign_volume(self):
         )
         wait_for_volume(
-            volume_manager=cinder_volumes,
+            volume_manager=cinder_client.volumes,
```
More changes related to the `cinder_api` factory function...see below.

wallrj commented on flocker/node/agents/test/blockdevicefactory.py:
```diff
@@ -179,7 +178,9 @@ def _openstack(**config):
     # bash (I already learned a piece of shell syntax today, once is all I can
     # take).
-    region = environ['FLOCKER_FUNCTIONAL_TEST_OPENSTACK_REGION'].upper()
+    region = environ.get('FLOCKER_FUNCTIONAL_TEST_OPENSTACK_REGION')
```
There's a bug in the Keystone Server / client API in our PistonCloud test environment which means that it doesn't respond well to region slugs.
This allows me to run the tests in the PistonCloud by omitting the region slug from the environment.

wallrj replied:
I should probably create an issue for that.

exarkun replied:
These changes should work on Rackspace too.  I'd confine this PR to the basic address-driven heuristic.  Then make fixes as necessary after there's a PistonCloud builder set up.

wallrj replied:
The PistonCloud builder now exists....so I think I'll leave this.

wallrj commented on flocker/node/agents/test/blockdevicefactory.py:
```diff
@@ -220,9 +221,9 @@ def _aws(**config):
 # factory.
 _BLOCKDEVICE_TYPES = {
-    ProviderType.openstack: (CinderBlockDeviceAPI, _openstack),
+    ProviderType.openstack: (cinder_api, _openstack),
```
This meant that none of the `LoggingNova???` wrappers were being used.

exarkun commented on flocker/common/_net.py:
```diff
@@ -0,0 +1,26 @@
+    """
+    Find all IPs for this machine.
+
+    :return: ``set`` of IP addresses (``bytes``).
```
An `ipaddr` type would make sense here (instead of `bytes`).

exarkun commented on flocker/node/agents/cinder.py:
```diff
@@ -169,36 +180,60 @@ def wait_for_volume(volume_manager, expected_volume,
+        to lists of address dictionaries in that network assigned to a
+        server.
+    :return: A ``set`` of all the IPv4 and IPv6 addresses from the
+        ``addresses`` attribute of a ``Server``.
```
`ipaddr` here too (currently bytes but the docstring doesn't say this)

exarkun commented on flocker/node/agents/cinder.py:
```diff
@@ -169,36 +180,60 @@ def wait_for_volume(volume_manager, expected_volume,
+        Find the Nova API server with the same non-loopback IP
+        addresses as this node.
+        That server is assumed to be this node.
+        Return the instance ID of that server.
```
Convert this to proper markup.  Also, what if no match is found?  The error case is probably worth handling explicitly.  We may also want to log enough information here that we could debug a failure caused by someone having a network configuration we didn't anticipate.

exarkun commented on flocker/node/agents/cinder.py:
```diff
@@ -291,8 +326,8 @@ def attach_volume(self, blockdevice_id, attach_to):
-            nova_volume = self.nova_volume_manager.get(blockdevice_id)
-        except NovaNotFound:
+            cinder_volume = self.cinder_volume_manager.get(blockdevice_id)
+        except CinderNotFound:
```
Hm.  Does this make any real difference?

If we stick with this, we no longer use `get` from the nova volume manager so maybe we should drop that from the interface.

exarkun replied:
(comment just below explains this change)

exarkun commented:
Thanks.  Looks good.  Please address inline comments (including your own :smile: ) and then merge.

wallrj commented:
I addressed the comments and forced a new build on my (unmerged) pistoncloud buildslave:
 http://23.253.99.92/builders/flocker%2Ffunctional%2Fpistoncloud%2Fcentos-7%2Fstorage-driver/builds/6

I'll merge tomorrow if all goes well.
